man/wl-restart.1.scd
wl-restart(1)

# NAME

wl-restart - restart your compositor when it crashes

# SYNOPSIS

*wl-restart* [[-h,-n N] --] <compositor args>

# DESCRIPTION

*wl-restart* is a simple tool that restarts your compositor when it crashes. It
also creates and destroys the Wayland socket for the compositor, so that
clients that support seamless compositor reconnects (most KDE programs, e.g.
*konsole*) don't die with it when your compositor crashes.

*wl-restart* will create a Wayland socket, and then starts the given compositor,
passing along the Wayland socket FD to the compositor with the options
*--socket NAME* and *--wayland-fd FD*. When the compositor crashes, *wl-restart*
will restart the compositor up to 10 times, after which it gives up. Details are
listed in *BEHAVIOUR* below.

Example Usage:

	$ wl-restart sway

	$ wl-restart -n 7 hyprland

This only works for compositors that support Wayland socket handover and that
support the *--socket* and *--wayland-fd* options or any of the other socket
handover mechanisms detailed under *OPTIONS*. Currently that is only Kwin and
Hyprland, but patches exist for sway and labwc. For details, see *COMPOSITOR
SUPPORT* below.

Currently there aren't many clients and toolkits that support surviving a
compositor restart, with *Qt6* and the KDE apps being a notable exception. For
details, see *CLIENT AND TOOLKIT SUPPORT* below.

# OPTIONS

*-h, --help*
	Show help message and exit.

*-n N, --max-restarts N*
	Restart the compositor a maximum of *N* times. (default is 10)

*--kde*
	Pass socket via cli options --socket and --wayland-fd. (default)
	This passes the socket in the same form as KDE's *kwin_wayland_wrapper*.

*--cli*
	Pass socket via cli options --wayland-socket and --wayland-fd.

*--env*
	Pass socket via env vars WAYLAND_SOCKET_NAME and WAYLAND_SOCKET_FD.

*--systemd*
	Pass socket via env vars LISTEN_PID, LISTEN_FDS, and LISTEN_FDNAMES.
	This passes the socket in the same form as the systemd socket activation
	protocol. Further information can be found in *sd_listen_fds*(3).

	The only difference from regular systemd socket activation is that in this
	case *wl-restart* will have already created the Wayland socket lock file,
	which would not be created by regular systemd socket activation as it is
	Wayland-specific.

# BEHAVIOUR

*wl-restart* first finds a free wayland socket name (e.g. *wayland-1*) and
creates it.

Then, it starts the compositor with the given arguments, but with *--socket
wayland-N --wayland-fd FD* appended to the arguments.

If the compositor exits with exit code *0*, *wl-restart* cleans up the socket
and exits as well.

If the compositor exits with exit code *133* (signal *SIGTRAP*), *wl-restart*
resets the restart counter and restarts the compositor.

If the compositor exits with any other exit code, *wl-restart* increments the
restart counter and restarts the compositor if it is smaller than the maximum.
Otherwise, *wl-restart* cleans up the socket and exits as well.

When the compositor is restarted, *wl-restart* sets the environment variable
*WL_RESTART_COUNT* to the restart counter, so other scripts can react to it and
e.g. display a message or restart services that don't restart automatically.

When *wl-restart* receives a *SIGINT* or *SIGTERM* signal, it kills the
compositor with *SIGTERM* if it's currently running, cleans up the socket, and
exits.

When *wl-restart* receives a *SIGHUP* signal, it kills the compositor with
*SIGTERM* if it's currently running, waits for it to die, and restarts it
without incrementing the restart counter.

# COMPOSITOR SUPPORT

*wl-restart* only works for compositors that support Wayland socket handover via
the *--socket* and *--wayland-fd* options.

*kwin_wayland*
	Kwin supports socket handover, but *wl-restart* isn't necessary or
	recommended for it, since automatic restarting for it is already implemented
	by *kwin_wayland_wrapper*, which is automatically used in
	*startplasma_wayland*.

	Kwin uses the *KDE* socket handover mechanism.

*hyprland*
	Hyprland supports socket handover and works well with *wl-restart*.

	Hyprland uses the *KDE* socket handover mechanism.

*sway*
	Upstream Sway does not support socket handover, but socket handover is
	implemented in the following branch and works well with *wl-restart*:

	https://github.com/ferdi265/sway/tree/feature-socket-handover

	This branch uses the *KDE* socket handover mechanism.
	
	https://github.com/ferdi265/sway/tree/feature-socket-handover-alternative

	This alternative branch uses the *env var* socket handover mechanism.

*labwc*
	Upstream labwc does not support socket handover, but socket handover is
	implemented in the following branch and works with *wl-restart*:

	https://github.com/ferdi265/labwc/tree/feature-socket-handover

	This branch uses the *env var* socket handover mechanism.

Other compositors do not support Wayland socket handover as of now. If there are
any compositors with support for socket handover missing from this list, please
open an issue for it on GitHub.

# CLIENT AND TOOLKIT SUPPORT

The socket handover implemented by *wl-restart* and compatible compositors
allows Wayland clients to survive a crash or restart of the compositor without
losing any state. However, this needs explicit support in the client or toolkit
used.

*Qt*
	Qt supports surviving a compositor restart in *Qt6*. The environment
	variable *QT_WAYLAND_RECONNECT=1* needs to be set in order for Qt apps to
	stay alive when the compositor crashes.

*GTK*
	GTK does not currently support surviving a compositor restart.
	There is a branch with a working implementation, but it hasn't been merged
	yet.

*SDL*
	SDL does not currently support surviving a compositor restart.
	There is a branch with a working implementation, but it hasn't been merged
	yet.
	
A more complete list of toolkits and their state supporting this can be found at
the following link:

	https://invent.kde.org/plasma/kwin/-/wikis/Restarting

# AUTHORS

Maintained by Ferdinand Bachmann <ferdinand.bachmann@yrlf.at>. More information
on *wl-restart* can be found at <https://github.com/Ferdi265/wl-restart>.


src/wl-restart.c
#define _GNU_SOURCE
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <stdio.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/wait.h>
#include <sys/signal.h>
#include <sys/errno.h>
#include "wl-socket.h"

// taken from libsystemd <systemd/sd-daemon.h>
// part of the systemd socket activation protocol
#define SD_LISTEN_FDS_START 3

typedef enum {
    MODE_KDE = 0,
    MODE_CLI = 1,
    MODE_ENV = 2,
    MODE_SYSTEMD = 3
} pass_mode_t;

typedef struct {
    struct wl_socket * socket;

    char ** compositor_argv;
    int compositor_argc;

    volatile pid_t compositor_pid;
    int restart_counter;
    int max_restarts;
    pass_mode_t mode;
} ctx_t;

void init(ctx_t * ctx) {
    ctx->socket = NULL;

    ctx->compositor_argv = NULL;
    ctx->compositor_argc = 0;

    ctx->compositor_pid = -1;
    ctx->restart_counter = 0;
    ctx->max_restarts = 10;
    ctx->mode = MODE_KDE;
}

void cleanup(ctx_t * ctx) {
    // close socket
    if (ctx->socket != NULL) {
        wl_socket_destroy(ctx->socket);
        ctx->socket = NULL;
    }

    // free argv
    if (ctx->compositor_argv != NULL) {
        for (int i = 0; i < ctx->compositor_argc; i++) {
            if (ctx->compositor_argv[i] != NULL) free(ctx->compositor_argv[i]);
        }

        free(ctx->compositor_argv);
        ctx->compositor_argv = NULL;
    }

    pid_t compositor_pid = ctx->compositor_pid;
    if (compositor_pid != -1) {
        ctx->compositor_pid = -1;
        kill(compositor_pid, SIGTERM);
    }
}

void exit_fail(ctx_t * ctx) {
    cleanup(ctx);
    exit(1);
}

void create_socket(ctx_t * ctx, int argc, char ** argv) {
    ctx->socket = wl_socket_create();
    if (ctx->socket == NULL) {
        fprintf(stderr, "error: failed to create wayland socket\n");
        exit_fail(ctx);
    }

    // allocate space for two extra options with arguments for cli modes
    ctx->compositor_argc = argc + 4;
    // add 1 for null terminator
    ctx->compositor_argv = calloc(ctx->compositor_argc + 1, sizeof (char *));

    // copy compositor arguments into argv array
    for (int i = 0; i < argc; i++) {
        ctx->compositor_argv[i] = strdup(argv[i]);
    }

    // prepare arguments for cli modes
    if (ctx->mode == MODE_KDE || ctx->mode == MODE_CLI) {
        const char * socket_name_arg = ctx->mode == MODE_KDE ? "--socket" : "--wayland-socket";
        const char * socket_fd_arg = "--wayland-fd";

        // add extra arguments
        char * socket_name = strdup(wl_socket_get_display_name(ctx->socket));
        ctx->compositor_argv[argc + 0] = strdup(socket_name_arg);
        ctx->compositor_argv[argc + 1] = socket_name;

        char * socket_fd = NULL;
        if (asprintf(&socket_fd, "%d", wl_socket_get_fd(ctx->socket)) == -1) {
            fprintf(stderr, "error: failed to convert fd to string\n");
            exit_fail(ctx);
        }

        ctx->compositor_argv[argc + 2] = strdup(socket_fd_arg);
        ctx->compositor_argv[argc + 3] = socket_fd;
    }

    // prepare environment for env modes
    if (ctx->mode == MODE_ENV || ctx->mode == MODE_SYSTEMD) {
        const char * socket_name_var = ctx->mode == MODE_ENV ? "WAYLAND_SOCKET_NAME" : "LISTEN_FDNAMES";
        const char * socket_fd_var = ctx->mode == MODE_ENV ? "WAYLAND_SOCKET_FD" : "LISTEN_FDS";

        // set environment vars
        setenv(socket_name_var, wl_socket_get_display_name(ctx->socket), true);

        if (ctx->mode == MODE_ENV) {
            char * socket_fd = NULL;
            if (asprintf(&socket_fd, "%d", wl_socket_get_fd(ctx->socket)) == -1) {
                fprintf(stderr, "error: failed to convert fd to string\n");
                exit_fail(ctx);
            }

            setenv(socket_fd_var, socket_fd, true);
            free(socket_fd);
        } else if (ctx->mode == MODE_SYSTEMD) {
            // systemd socket activation has a hardcoded fd number
            // and only passes the number of fds
            setenv(socket_fd_var, "1", true);
        }
    }
}

void start_compositor(ctx_t * ctx) {
    pid_t pid = fork();
    if (pid == 0) {
        if (ctx->mode == MODE_SYSTEMD) {
            // set up systemd socket activation

            // set LISTEN_PID env var
            char * listen_pid = NULL;
            if (asprintf(&listen_pid, "%d", getpid()) == -1) {
                fprintf(stderr, "error: failed to convert compositor pid to string\n");
                exit(1);
            }
            setenv("LISTEN_PID", listen_pid, true);
            free(listen_pid);

            // move socket fd to SD_LISTEN_FDS_START
            int socket_fd = wl_socket_get_fd(ctx->socket);
            if (socket_fd != SD_LISTEN_FDS_START) {
                if (dup2(socket_fd, SD_LISTEN_FDS_START) == -1) {
                    fprintf(stderr, "error: failed to move socket fd to SD_LISTEN_FDS_START\n");
                    exit(1);
                }

                // close original fd so it doesn't leak
                close(socket_fd);
            }
        }

        // exec into compositor process
        execvp(ctx->compositor_argv[0], ctx->compositor_argv);

        fprintf(stderr, "error: failed to start compositor\n");
        exit(1);
    } else {
        ctx->compositor_pid = pid;
    }
}

void wait_compositor(ctx_t * ctx, int * exit_status, int * exit_signal) {
    int stat;
    pid_t compositor_pid = ctx->compositor_pid;
    while (compositor_pid != -1 && waitpid(compositor_pid, &stat, 0) == -1) {
        if (errno == EINTR) {
            // interrupted by signal
            // try again
        } else if (errno == ECHILD) {
            // signal handler probably already waited for this child
            // try again
        } else {
            fprintf(stderr, "error: failed to wait for compositor: %s\n", strerror(errno));
            exit_fail(ctx);
        }

        // re-check compositor pid
        // signal handler might have already killed compositor
        compositor_pid = ctx->compositor_pid;
    }

    if (ctx->compositor_pid == -1) {
        // signal handler already killed the compositor
        // simulate SIGHUP
        *exit_status = -1;
        *exit_signal = SIGHUP;
    } else if (WIFSIGNALED(stat)) {
        // compositor killed by signal
        *exit_status = -1;
        *exit_signal = WTERMSIG(stat);
    } else {
        // compositor died normally
        *exit_status = WEXITSTATUS(stat);
        *exit_signal = -1;
    }

    // clear compositor pid
    ctx->compositor_pid = -1;
}

static ctx_t * signal_ctx = NULL;
void handle_quit_signal(int signal) {
    ctx_t * ctx = signal_ctx;

    fprintf(stderr, "info: signal %s received, quitting\n", strsignal(signal));
    exit_fail(ctx);
}

void handle_restart_signal(int signal) {
    ctx_t * ctx = signal_ctx;

    fprintf(stderr, "info: signal %s received, restarting compositor\n", strsignal(signal));
    pid_t compositor_pid = ctx->compositor_pid;
    if (compositor_pid != -1) {
        ctx->compositor_pid = -1;
        kill(compositor_pid, SIGTERM);
    }
}

void register_signals(ctx_t * ctx) {
    signal_ctx = ctx;
    signal(SIGINT, handle_quit_signal);
    signal(SIGTERM, handle_quit_signal);
    signal(SIGHUP, handle_restart_signal);
}

void run(ctx_t * ctx) {
    while (ctx->restart_counter < ctx->max_restarts) {
        start_compositor(ctx);
        int exit_status = -1;
        int exit_signal = -1;
        wait_compositor(ctx, &exit_status, &exit_signal);

        if (exit_status == 0) {
            fprintf(stderr, "info: compositor exited successfully, quitting\n");
            cleanup(ctx);
            exit(0);
        } else if (exit_status > 0) {
            ctx->restart_counter++;
            fprintf(stderr, "info: compositor exited with code %d, incrementing restart counter (%d)\n", exit_status, ctx->restart_counter);
        } else if (exit_signal == SIGHUP) {
            ctx->restart_counter = 0;
            fprintf(stderr, "info: compositor died with signal %s, restarting\n", strsignal(SIGHUP));
        } else if (exit_signal == SIGTRAP) {
            ctx->restart_counter = 0;
            fprintf(stderr, "info: compositor died with signal %s, resetting counter\n", strsignal(SIGTRAP));
        } else if (exit_signal > 0) {
            ctx->restart_counter++;
            fprintf(stderr, "info: compositor died with signal %s, incrementing restart counter (%d)\n", strsignal(exit_signal), ctx->restart_counter);
        } else {
            fprintf(stderr, "error: failed to detect how compositor exited\n");
            exit_fail(ctx);
        }

        // format new restart count
        char * restart_counter_str = NULL;
        if (asprintf(&restart_counter_str, "%d", ctx->restart_counter) == -1) {
            fprintf(stderr, "error: failed to convert restart counter to string\n");
            exit_fail(ctx);
        }

        // add restart count to environment
        setenv("WL_RESTART_COUNT", restart_counter_str, true);
        free(restart_counter_str);
    }

    fprintf(stderr, "error: too many restarts, quitting\n");
    exit_fail(ctx);
}

void usage(ctx_t * ctx) {
    printf("usage: wl-restart [[options] --] <compositor args>\n");
    printf("\n");
    printf("compositor restart helper. restarts your compositor when it\n");
    printf("crashes and keeps the wayland socket alive.\n");
    printf("\n");
    printf("options:\n");
    printf("  -h,   --help           show this help\n");
    printf("  -n N, --max-restarts N restart a maximum of N times (default 10)\n");
    printf("        --kde            pass socket via cli options --socket and --wayland-fd (default)\n");
    printf("        --cli            pass socket via cli options --wayland-socket and --wayland-fd\n");
    printf("        --env            pass socket via env vars WAYLAND_SOCKET_NAME and WAYLAND_SOCKET_FD\n");
    printf("        --systemd        pass socket via env vars LISTEN_PID, LISTEN_FDS, and LISTEN_FDNAMES\n");
    cleanup(ctx);
    exit(0);
}

int main(int argc, char ** argv) {
    if (argc > 0) {
        // skip program name
        argc--, argv++;
    }

    ctx_t ctx;
    init(&ctx);

    while (argv[0] != NULL && argv[0][0] == '-') {
        char * opt = argv[0];
        char * arg = argv[1];
        argc--, argv++;

        if (strcmp(opt, "-h") == 0 || strcmp(opt, "--help") == 0) {
            usage(&ctx);
        } else if (strcmp(opt, "-n") == 0 || strcmp(opt, "--max-restarts") == 0) {
            if (arg == NULL) {
                fprintf(stderr, "error: option '%s' needs an argument\n", opt);
                exit_fail(&ctx);
            }

            argc--, argv++;
            ctx.max_restarts = atoi(arg);
        } else if (strcmp(opt, "--kde") == 0) {
            ctx.mode = MODE_KDE;
        } else if (strcmp(opt, "--cli") == 0) {
            ctx.mode = MODE_CLI;
        } else if (strcmp(opt, "--env") == 0) {
            ctx.mode = MODE_ENV;
        } else if (strcmp(opt, "--systemd") == 0) {
            ctx.mode = MODE_SYSTEMD;
        } else if (strcmp(opt, "--") == 0) {
            break;
        } else {
            fprintf(stderr, "error: unknown option '%s', see --help\n", opt);
            exit_fail(&ctx);
        }
    }

    if (argc == 0) {
        usage(&ctx);
    } else {
        create_socket(&ctx, argc, argv);
        register_signals(&ctx);
        run(&ctx);
    }

    cleanup(&ctx);
}
